libs/remix-lib/src/execution/txRunnerWeb3.ts

```diff
@@ -0,0 +1,156 @@
+      return this.getWeb3().eth.call(tx, function (error, result: any) {
+        if (error) return callback(error)
+        callback(null, {
+          result: result
+        })
+      })
+    }
+    this.getWeb3().eth.estimateGas(tx, (err, gasEstimation) => {
+      if (err && err.message.indexOf('Invalid JSON RPC response') !== -1) {
+        // // @todo(#378) this should be removed when https://github.com/WalletConnect/walletconnect-monorepo/issues/334 is fixed
+        callback(new Error('Gas estimation failed because of an unknown internal error. This may indicated that the transaction will fail.'))
+      }
+      gasEstimationForceSend(err, () => {
+        // callback is called whenever no error
+        tx['gas'] = !gasEstimation ? gasLimit : gasEstimation
+
+        if (this._api.config.getUnpersistedProperty('doNotShowTransactionConfirmationAgain')) {
+          return this._executeTx(tx, null, this._api, promptCb, callback)
+        }
+
+        this._api.detectNetwork((err, network) => {
+          if (err) {
+            console.log(err)
+            return
+          }
+
+          confirmCb(network, tx, tx['gas'], (gasPrice) => {
+            return this._executeTx(tx, gasPrice, this._api, promptCb, callback)
+          }, (error) => {
+            callback(error)
+          })
+        })
+      }, () => {
+        const blockGasLimit = this.currentblockGasLimit()
+        // NOTE: estimateGas very likely will return a large limit if execution of the code failed
+        //       we want to be able to run the code in order to debug and find the cause for the failure
+        if (err) return callback(err)
+
+        let warnEstimation = ' An important gas estimation might also be the sign of a problem in the contract code. Please check loops and be sure you did not sent value to a non payable function (that\'s also the reason of strong gas estimation). '
+        warnEstimation += ' ' + err
+
+        if (gasEstimation > gasLimit) {
+          return callback('Gas required exceeds limit: ' + gasLimit + '. ' + warnEstimation)
+        }
+        if (gasEstimation > blockGasLimit) {
+          return callback('Gas required exceeds block gas limit: ' + gasLimit + '. ' + warnEstimation)
+        }
+      })
+    })
+  }
+}
+
+async function tryTillReceiptAvailable (txhash, web3) {
+  return new Promise((resolve, reject) => {
+    web3.eth.getTransactionReceipt(txhash, async (err, receipt) => {
+      if (err || !receipt) {
+        // Try again with a bit of delay if error or if result still null
+        await pause()
+        return resolve(await tryTillReceiptAvailable(txhash, web3))
+      }
+      return resolve(receipt)
+    })
+  })
+}
+
+async function tryTillTxAvailable (txhash, web3) {
+  return new Promise((resolve, reject) => {
+    web3.eth.getTransaction(txhash, async (err, tx) => {
+      if (err || !tx) {
+        // Try again with a bit of delay if error or if result still null
+        await pause()
```

Comment:
can you have an infinite loop here?

Reply:
That 's not blocking the UI but yes, this will never stop verifying the hash.
We can have a reject after waiting for more than x minutes.
This PR is not updating this behavior, I would address that isfnecessary in a subsequent PR.